nifi-extension-bundles/nifi-parquet-bundle/nifi-parquet-content-viewer/src/main/java/org/apache/parquet/web/controller/ParquetContentViewerController.java

```diff
@@ -0,0 +1,155 @@
+import java.io.InputStream;
+import java.net.HttpURLConnection;
+
+public class ParquetContentViewerController extends HttpServlet {
+    private static final Logger logger = LoggerFactory.getLogger(ParquetContentViewerController.class);
+    private static final long maxBytes = 1024 * 1024 * 2; //10MB
+    private static final int bufferSize = 8 * 1024; // 8KB
+    private static final ObjectMapper objectMapper = new ObjectMapper().enable(SerializationFeature.INDENT_OUTPUT);
+
+    @Override
+    public void doGet(final HttpServletRequest request, final HttpServletResponse response) throws IOException {
+        final ContentRequestContext requestContext = new HttpServletContentRequestContext(request);
+
+        final ServletContext servletContext = request.getServletContext();
+        final ContentAccess contentAccess = (ContentAccess) servletContext.getAttribute("nifi-content-access");
+        // get the content
+        final DownloadableContent downloadableContent;
+        try {
+            downloadableContent = contentAccess.getContent(requestContext);
+        } catch (final ResourceNotFoundException e) {
+            logger.warn("Content not found", e);
+            response.sendError(HttpURLConnection.HTTP_NOT_FOUND, "Content not found");
+            return;
+        } catch (final AccessDeniedException e) {
+            logger.warn("Content access denied", e);
+            response.sendError(HttpURLConnection.HTTP_FORBIDDEN, "Content access denied");
+            return;
+        } catch (final Exception e) {
+            logger.warn("Content retrieval failed", e);
+            response.sendError(HttpURLConnection.HTTP_INTERNAL_ERROR, "Content retrieval failed");
+            return;
+        }
+
+        response.setStatus(HttpServletResponse.SC_OK);
+
+        final boolean formatted = Boolean.parseBoolean(request.getParameter("formatted"));
+        if (!formatted) {
+            final InputStream contentStream = downloadableContent.getContent();
+            contentStream.transferTo(response.getOutputStream());
+            return;
+        }
+
+        // allow the user to drive the data type but fall back to the content type if necessary
+        String displayName = request.getParameter("mimeTypeDisplayName");
+        if (displayName == null) {
+            displayName = downloadableContent.getType();
+        }
+
+        if (displayName == null || !(displayName.equals("parquet") || displayName.equals("application/vnd.apache.parquet"))) {
+            response.sendError(HttpURLConnection.HTTP_BAD_REQUEST, "Unknown content type");
+            return;
+        }
+
+        try {
+            //Convert InputStream to a seekable InputStream
+            byte[] data = getInputStreamBytes(downloadableContent.getContent());
+
+            if (data.length == 0) {
+                response.getOutputStream().write("Content size is too large to display.".getBytes());
+                return;
+            }
```

Comment:
This should set a different HTTP response code, probably an Internal Server Error.